myhoard/restore_coordinator.py

```diff
@@ -430,6 +436,66 @@ def restore_basebackup(self) -> None:
         finally:
             self.basebackup_restore_operation = None
 
+    def rebuild_tables(self) -> None:
+        self._ensure_mysql_server_is_started(with_binlog=False, with_gtids=False)
+
+        def escape_identifier(name: str) -> str:
+            return "`" + name.replace("`", "``") + "`"
+
+        with self._mysql_cursor() as cursor:
+            cursor.execute(
+                "SELECT TABLE_SCHEMA,TABLE_NAME,TABLE_ROWS,AVG_ROW_LENGTH"
+                " FROM INFORMATION_SCHEMA.TABLES WHERE ENGINE='InnoDB'"
+            )
+            tables = cursor.fetchall()
+            tables = sorted(
+                [
+                    table
+                    for table in tables
+                    if (table["TABLE_SCHEMA"], table["TABLE_NAME"])
+                    not in {
+                        ("mysql", "innodb_index_stats"),
+                        ("mysql", "innodb_table_stats"),
+                    }
+                ],
+                key=lambda table: (table["TABLE_SCHEMA"], table["TABLE_NAME"]),
+            )
+            estimated_total_bytes = sum(table["TABLE_ROWS"] * table["AVG_ROW_LENGTH"] for table in tables)
+            estimated_progress_bytes = 0
+            self.log.info("Will rebuild %s tables, estimated total size: %s bytes", len(tables), estimated_total_bytes)
+            for table_index, table in enumerate(tables, start=1):
```

Comment:
Might be good to split this processing so that there's separate step for listing the tables that need to be rebuilt and those are stored in state as well as the latest table that has been built thus far so that operation resumes from where it was left if it's aborted instead of starting over from beginning.

Reply:
I didn't really need to save the table list to be able to resume.

Do you think about storing the list of tables for some specific reason or just as a possible implementation ?

Reply:
Current implementation should be fine. I was thinking of not rerunning the query to list tables but refreshing those might be good idea and some pathological services could have extremely large number of tables so avoiding storing those in state makes sense.